scipy/stats/_morestats.py

```diff
@@ -4047,12 +4058,16 @@ def directionalmean(samples, *, axis=0, normalize=True):
 
     Returns
     -------
-    directionalmean : ndarray
+    mean_direction : ndarray
         Directional mean.
+    mean_resultant_length : ndarray
+        length of the mean resultant vector. This can be used to calculate
+        the directional variance.
```

Comment:
Is this the correct docstring for returning a NamedTuple? I have never used them actually. But like this I would assume that two arrays are returned.

Reply:
It has been done like this a lot in the past, but recently it's been done like for [`kstest`](http://scipy.github.io/devdocs/reference/generated/scipy.stats.kstest.html#scipy.stats.kstest) for example. There is one return object described as "an object containing the following attributes". We can also mention that it can be unpacked as a tuple, but I think we've neglected that in the recent past, preferring that the attributes should be used. But that is less important when there are only two attributes, in which case the order is easier to remember.

Reply:
Are we drifting away from the original reason for introducing the "tuple bunch"?  The lesson learned from the old `stats` functions is that returning a tuple (named or not) is a brittle API.  Adding more information to the returned object is hard to do in a clean and backwards compatible way.  *New* functions should just return a class instance with attributes that hold the computed values, with no need for a tuple-like interface.

Reply:
> Adding more information to the returned object is hard to do in a clean and backwards compatible way.

It's quite easy, really. We just swap out `namedtuple` for `make_tuple_bunch`.

I make classes for new code, personally. But since the problem of extending tuples is solved, I don't think of there being very strict rules about the return values. 

As for advertising the tuple behavior, I'd prefer that we didn't. As I mentioned, I haven't advertised that behavior in the past. I'll strikethrough my previous comment about that to clarify.

Reply:
`make_tuple_bunch` was created as a way to allow the return values of existing functions that returned a tuple to be extended with new attributes while preserving the old tuple-like behavior.  For *new* functions, there is no need for `make_tuple_bunch`.  Just use a class.

Reply:
> But since the problem of extending tuples is solved, I don't think of there being very strict rules about needing to do so.

The problem is that people will stumble upon the tuple nature of the return object and start using tuple unpacking, even if you don't advertise it. We don't want people to do that. We want people to use the single advertised interface which should be an object with attributes. Please don't use `make_tuple_bunch` for new code.

Reply:
We're not.

Reply:
Just for consistency, I kept it namedtuple but if someone suggests otherwise, I am OK to change it to a dataclass.

Reply:
> If you mean not to use namedtuples, feel free to propose that in a separate issue.

We already had that discussion. [Here's you applying that decision in another issue.](https://github.com/scipy/scipy/pull/11119#issuecomment-681134869) We no longer want to use `tuple`s, `namedtuple`s, or `make_tuple_bunch` in new functions. For new functions with multiple return values, we make classes with only attribute access and no tuple semantics.

Reply:
Thanks for pointing to the issue @rkern! I have changed the namedtuple to a class now. Let me know if you have anymore comments in this @mdhaber @rkern!

Reply:
Hmm. I'm not really seeing it there. Anyway, since we've all been in agreement from the start about what _should_ happen, I've created gh-17306 so it doesn't need to be discussed in the future.